src/plugins/hetero/src/compiled_model.cpp

```diff
@@ -0,0 +1,708 @@
+        // Get all subgraph inputs using just node affinities. Also collect transitive closure
+        for (auto&& node : orderedOps) {
+            if (ov::op::util::is_parameter(node) || ov::op::util::is_constant(node)) {
+                graphInputNodes.insert(node);
+                subgraphInputs.insert(Input{node.get(), 0});
+                nodeInputDependencies[node].insert(Input{node.get(), 0});
+            } else {
+                auto inputs = node->inputs();
+                auto& nodeInputDependency = nodeInputDependencies[node];
+                for (auto&& input : inputs) {
+                    nodeInputDependency.insert(input);
+                    auto& inputDependency = nodeInputDependencies[InputNode(input)];
+                    nodeInputDependency.insert(inputDependency.begin(), inputDependency.end());
+                    if (affinities[node] != affinities[InputNode(input)]) {
+                        subgraphInputs.insert(input);
+                    }
+                }
+            }
+        }
+
+        // Assign each node subgraph ID
+        auto CollectSubgraphs = [&] {
+            std::deque<int> subgraphIds;
+            NodeMap<int*> subgraphIdPtrs;
+            for (auto&& node : orderedOps) {
+                auto allNodeInputs = node->inputs();
+                std::vector<Input> inputs;
+                for (auto&& input : allNodeInputs) {
+                    if (!ov::util::contains(subgraphInputs, input)) {
+                        inputs.emplace_back(std::move(input));
+                    }
+                }
+                if (inputs.empty()) {
+                    subgraphIds.push_back(static_cast<int>(subgraphIds.size()));
+                    subgraphIdPtrs.emplace(node, &(subgraphIds.back()));
+                } else {
+                    auto firstInputSubgraphIdPtr = subgraphIdPtrs[InputNode(inputs.front())];
+                    for (auto&& input : inputs) {
+                        auto inputId = *subgraphIdPtrs[InputNode(input)];
+                        for (auto& subgraphId : subgraphIds) {
+                            if (subgraphId == inputId) {
+                                subgraphId = *firstInputSubgraphIdPtr;
+                            }
+                        }
+                    }
+                    subgraphIdPtrs.emplace(node, firstInputSubgraphIdPtr);
+                }
+            }
+            NodeMap<int> result;
+            for (auto&& subgraphIdPtr : subgraphIdPtrs) {
+                result.emplace(subgraphIdPtr.first, *(subgraphIdPtr.second));
+            }
+            return result;
+        };
+
+        // Split cyclic dependencies.
+        for (std::size_t prevSubgraphs = 0, cyclicSplitStep = 0; prevSubgraphs != subgraphInputs.size();
+             ++cyclicSplitStep) {
+            OPENVINO_ASSERT(cyclicSplitStep < orderedOps.size());
+            prevSubgraphs = subgraphInputs.size();
+            auto subgraphIds = CollectSubgraphs();
```

Comment:
Can we call `CollectSubgraphs()` once?
```suggestion
        auto subgraphIds = CollectSubgraphs();
        // Split cyclic dependencies.
        for (std::size_t prevSubgraphs = 0, cyclicSplitStep = 0; prevSubgraphs != subgraphInputs.size();
             ++cyclicSplitStep) {
            OPENVINO_ASSERT(cyclicSplitStep < orderedOps.size());
            prevSubgraphs = subgraphInputs.size();
```

Reply:
> Can we call `CollectSubgraphs()` once?

No, subgraphs are recalculated several times during this loop. New splits are inserted. Let's keep this part untouch in this PR. It is not connected to API 2.0